src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -0,0 +1,756 @@
+	AddTag(name, value string)
+	SamplesAppender() (SamplesAppender, error)
+	Reset()
+	Finalize()
+}
+
+// SamplesAppender is a downsampling samples appender,
+// that can only be called by a single caller at a time.
+type SamplesAppender interface {
+	AppendCounterSample(value int64) error
+	AppendGaugeSample(value float64) error
+}
+
+// DownsamplerOptions is a set of required downsampler options.
+type DownsamplerOptions struct {
+	Storage                 storage.Storage
+	StorageFlushConcurrency int
+	RulesKVStore            kv.Store
+	ClockOptions            clock.Options
+	InstrumentOptions       instrument.Options
+	TagEncoderOptions       serialize.TagEncoderOptions
+	TagDecoderOptions       serialize.TagDecoderOptions
+	TagEncoderPoolOptions   pool.ObjectPoolOptions
+	TagDecoderPoolOptions   pool.ObjectPoolOptions
+}
+
+// Validate will validate the dynamic downsampling options.
+func (o DownsamplerOptions) Validate() error {
+	if o.Storage == nil {
+		return errNoStorage
+	}
+	if o.RulesKVStore == nil {
+		return errNoRulesStore
+	}
+	if o.ClockOptions == nil {
+		return errNoClockOptions
+	}
+	if o.InstrumentOptions == nil {
+		return errNoInstrumentOptions
+	}
+	if o.TagEncoderOptions == nil {
+		return errNoTagEncoderOptions
+	}
+	if o.TagDecoderOptions == nil {
+		return errNoTagDecoderOptions
+	}
+	if o.TagEncoderPoolOptions == nil {
+		return errNoTagEncoderPoolOptions
+	}
+	if o.TagDecoderPoolOptions == nil {
+		return errNoTagDecoderPoolOptions
+	}
+	return nil
+}
+
+type newAggregatorResult struct {
+	aggregator aggregator.Aggregator
+
+	clockOpts clock.Options
+
+	matcher matcher.Matcher
+
+	tagEncoderPool serialize.TagEncoderPool
+	tagDecoderPool serialize.TagDecoderPool
+
+	encodedTagsIteratorPool *encodedTagsIteratorPool
+}
+
+func (o DownsamplerOptions) newAggregator() (newAggregatorResult, error) {
+	// Validate options first.
+	if err := o.Validate(); err != nil {
+		return newAggregatorResult{}, err
+	}
+
+	var (
+		storageFlushConcurrency = defaultStorageFlushConcurrency
+		rulesStore              = o.RulesKVStore
+		clockOpts               = o.ClockOptions
+		instrumentOpts          = o.InstrumentOptions
+	)
+	if o.StorageFlushConcurrency > 0 {
+		storageFlushConcurrency = storageFlushConcurrency
+	}
+
+	// Configure rules options.
+	tagEncoderPool := serialize.NewTagEncoderPool(o.TagEncoderOptions,
+		o.TagEncoderPoolOptions)
+	tagEncoderPool.Init()
+
+	tagDecoderPool := serialize.NewTagDecoderPool(o.TagDecoderOptions,
+		o.TagDecoderPoolOptions)
+	tagDecoderPool.Init()
+
+	sortedTagIteratorPool := newEncodedTagsIteratorPool(tagDecoderPool,
+		o.TagDecoderPoolOptions)
+	sortedTagIteratorPool.Init()
+
+	sortedTagIteratorFn := func(tagPairs []byte) id.SortedTagIterator {
+		it := sortedTagIteratorPool.Get()
+		it.Reset(tagPairs)
+		return it
+	}
+
+	tagsFilterOptions := filters.TagsFilterOptions{
+		NameTagKey: metricNameTagName,
+		NameAndTagsFn: func(id []byte) ([]byte, []byte, error) {
```

Comment:
Does this mean for prom metrics name is a separate tag just like any other normal tags?

Reply:
Yes indeed.